proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1569 @@
+  spec:
+    #@schema/validation one_of=["aws", "azure", "vsphere"]
+    provider: ""
+```
+
+#### Use Case: Union Structures
+
+An author can require that exactly one of a set of sibling node has a value:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  #@schema/validation one_not_null=["oidc", "ldap"]
+  config:
+    #@schema/nullable
+    oidc:
+      CLIENT_ID: null #! required if oidc enabled
+      CLIENT_SECRET: null #! required if oidc enabled
+      issuer: null #! <OIDC_IDP_URL> is required if oidc enabled
+    #@schema/nullable
+    ldap:
+      host: null #! <LDAP_HOST> is required if ldap enabed
+      bindDN: null
+      bindPW: null
+```
+
+This approach is simpler than (and therefore recommended over) using a discriminator:
+
+```yaml
+#@data/values
+---
+dex:
+  config:
+    type: "oidc"
+    oidc:
+      CLIENT_ID: null #! required if oidc enabled
+      CLIENT_SECRET: null #! required if oidc enabled
+      issuer: null #! <OIDC_IDP_URL> is required if oidc enabled
+    ldap:
+      host: null #! <LDAP_HOST> is required if ldap enabed
+      bindDN: null
+      bindPW: null
+```
+
+#### Use Case: Multiple Nodes
+
+When a validation involves more than just one Node's value:
+
+```python
+def validate_registry():
+ if data.values.persistence.imageChartStorage.type == "filesystem":
+   if data.values.persistence.persistentVolumeClaim.registry.accessMode == "ReadWriteOnce":
+     data.values.registry.replicas == 1 or assert.fail("The registry replicas must be 1 when the image storage is filesystem and the access mode of persistentVolumeClaim is ReadWriteOnce")
+   end
+ end
+end
+```
+(ref: [vmware-tanzu/community-edition/.../harbor/2.2.3/config/values.star](https://github.com/vmware-tanzu/community-edition/blob/cc7eb24b0e13875a06ac8578cd73f83a217ec4d9/addons/packages/harbor/2.2.3/bundle/config/values.star#L45-L53))
+
+The author places the validation at the lowest-common parent Node:
+
+```yaml
+#@ def one_registry_if_pvc_is_filesystem(val):
+#@   if val.persistence.imageChartStorage.type == "filesystem" and
+#@      val.persistence.persistentVolumeClaim.registry.accessMode == "ReadWriteOnce":
+#@      return val.registry.replicas == 1 \
+#@          or (False, "There can be only one registry replica when the image storage is filesystem and the access mode of persistentVolumeClaim is ReadWriteOnce; there are {} replicas.".format(val.registry.replicas))
+#@   end
+#@ end
+
+#@data/values-schema
+#@schema/validation ("There should be exactly one (1) registry replica if Helm Charts are stored on the filesystem.", one_registry_if_pvc_is_filesystem)
+---
+persistence:
+  imageChartStorage:
+    type: filesystem
+  persistentVolumeClaim:
+    registry:
+      accessMode: ReadWriteOnce
+registry:
+  replicas: 2
+```
+
+Note: the `@ytt:data` module's `data.values` struct is _not_ populated until _after_ the Data Values Pre-Processing phase is complete.  As such, the `data.values` variable is not useful within validations.
```

Comment:
This is an interesting point! I was assuming `data.values` would be available.

Reply:
I go back and forth on this one.

I see the convenience of being able to reference _any_ data value from within a validation rule.

However, doing so introduces some subtle accidental complexity:
- the validation rule makes assumptions about the shape of data values (whereas a rule defined on a node only makes assumptions about the sub-tree on which it annotates)
- data values are _being validated_ when validation rules are running — attempts to make for a sane reliable output (e.g. traversing the data values tree from the leaves, up) can easily be foiled if any rule has access to any node and fails because that _other_ value is invalid.

If we can make those situations better somehow, we can relax this constraint.

In the meantime, it would be friendly/helpful if we were clear about this constraint in the documentation for this feature. 👍🏻 

btw: still open to other perspectives... would love to hear if you feel strongly the other way.

Reply:
That is _exactly_ the intended approach.